raster-foundry/raster-foundry
# Add Dao and Routes for Annotation Project and its related

## Overview

This PR adds the below endpoints along with its required Dao methods and some data models:
- [x] GET `/annotation-projects/`
- [x] GET `/annotation-projects/<id>` (should include `label class groups` and `label classes` for this project as related info)
  - [x] also include classes/label summaries
  - [x] also include also include summary of task statuses
- [x] GET `/annotation-projects/<id>/tasks`
- [x] GET `/annotation-projects/<id>/tasks/user-stats` -- replacement for summary endpoint used today
- [x] POST `/annotation-projects/<id>/tasks/<id>/labels`
- [x] POST `/annotation-projects/<id>/tasks/<id>/lock`
- [x] DELETE `/annotation-projects/<id>/tasks/<id>/lock`
- [x] DELETE `/annotation-projects/<id>`
- [x] PUT `/annotation-projects/<id>`
- [x] [permission routes](https://github.com/raster-foundry/raster-foundry/blob/develop/app-backend/api/src/main/scala/project/Routes.scala#L475-L490)

### Checklist

- ~Swagger specification updated~
- ~New tables and queries have appropriate indices added~
- ~Any content changes are properly templated using `BUILDCONFIG.APP_NAME`~
- ~Any new SQL strings have tests~ Not done, but will be in https://github.com/raster-foundry/raster-foundry/issues/5302

### Notes

Some endpoints give `403`s rn, and probably due to some scope thing that I can't figure out. Will need to pair with @jisantuc on Monday to look at them.

Property tests are not added along with new sql, but will be in https://github.com/raster-foundry/raster-foundry/issues/5302

## Testing Instructions

- Run migrations
- If you have not run batch job to populate the `annotation_projects` table, it should be run here too
- Give your local user `groundworkUser` scope in DB
- assemble api jar and spin up servers
- make sure your user is in the labeler team and validator team in your local DB
- follow https://github.com/raster-foundry/raster-foundry/pull/5294 to create some annotation projects to prepare for later steps
- test out the above listed endpoints:
   * GET `/annotation-projects/` should return paginated annotation projects in your DB, each returned result should also include label class groups with label classes, and include tile layers, if there are any. The following QPs should also work:
        * `search` should do a name search
        * `ownershipType` should be one of: not present, or `owned`, or `shared`, or `inherited`, which will be used on frontend
        * `projectType ` should be one of: `detection`, `classification`, `segmentation`
        * `groupType ` should be one of: `platform`, `organization`, `team`
        * `groupId` should be an UUID of a plat/org/team
   * PUT `/annotation-projects/<id>` with an updated `AnnotationProject.WithRelated` object should work
   * DELETE `/annotation-projects/<id>` to delete one of the projects you created in the early step
   * GET `/annotation-projects/<id>` should show a project's detail with task status summary, and label count summary, for example, see this for the Florence Alabama annotation project
   * GET `/annotation-projects/<id>/tasks` should list paginated tasks of this annotation projects
   * GET `/annotation-projects/<id>/tasks/user-stats` should return a summary of user tasks
   * POST `/annotation-projects/<id>/tasks/<id>/lock` should lock a task
   * DELETE `/annotation-projects/<id>/tasks/<id>/lock` should remove the lock on this task
   * POST `/annotation-projects/<id>/tasks/<id>/labels` with an object of `AnnotationLabelWithClassesFeatureCollectionCreate` should work

Closes https://github.com/raster-foundry/raster-foundry/issues/5282
